test/langtools/tools/javac/records/RecordsBinaryCompatibilityTests.java

```diff
@@ -0,0 +1,307 @@
+                public record R(int i, String j) {}
+                """,
+                """
+                package pkg;
+                public record R(String j, int i) {}
+                """,
+                """
+                package pkg;
+                public class Client {
+                    public static void main(String... args) {
+                        R r = new R(0, "");
+                        System.out.println(r.i());
+                    }
+                }
+                """,
+                true,
+                "java.lang.NoSuchMethodError"
+        );
+    }
+
+    @Test
+    public void testCompatibilityAfterChangingRecordComponent2(Path base) throws Exception {
+        testCompatibilityAfterModifyingRecord(
+                base,
+                """
+                package pkg;
+                public record R(int j) {
+                    public static int i() { return 0; }
+                }
+                """,
+                """
+                package pkg;
+                public record R(int i) {}
+                """,
+                """
+                package pkg;
+                public class Client {
+                    public static void main(String... args) {
+                        R r = new R(0);
+                        System.out.println(r.i());
+                    }
+                }
+                """,
+                true,
+                "java.lang.IncompatibleClassChangeError"
+        );
+    }
+
+    @Test
+    public void testCompatibilityAfterChangingRecordComponent3(Path base) throws Exception {
+        testCompatibilityAfterModifyingRecord(
+                base,
+                """
+                package pkg;
+                public record R(int i) {
+                }
+                """,
+                """
+                package pkg;
+                public record R(int j) {}
+                """,
+                """
+                package pkg;
+                public class Client {
+                    public static void main(String... args) {
+                        R r = new R(0);
+                        System.out.println(r.i());
+                    }
+                }
+                """,
+                true,
+                "java.lang.NoSuchMethodError"
+        );
+    }
+
+    /* 1- compiles the first version of the record class source code along with the client source code
+     * 2- executes the client class just to make sure that it works
+     * 3- compiles the second version of the record class
+     * 4- executes the client class and makes sure that the VM throws the expected error or not
+     *    depending on the shouldFail argument
+     */
+    private void testCompatibilityAfterModifyingRecord(
+            Path base,
+            String recordCode1,
+            String recordCode2,
+            String clientCode,
+            boolean shouldFail,
+            String expectedError) throws Exception {
```

Comment:
Since the expected error always seems to be the name of a class, you could make the type of the parameter be `Class<?>` and then pass class literals as arguments. Inside the method, you get use `getName()` to get the name you expect to see.

Reply:
I have uploaded a new iteration addressing your comments, thanks